sd-finetuner-workflow/sd-finetuner/finetuner.py

```diff
@@ -0,0 +1,488 @@
+            optimizer_cls = torch.optim.AdamW
+    else:
+        optimizer_cls = torch.optim.AdamW
+    
+    optimizer = optimizer_cls(
+        unet.parameters(),
+        lr=args.lr,
+        betas=(args.adam_beta1, args.adam_beta2),
+        eps=args.adam_epsilon,
+        weight_decay=args.adam_weight_decay,
+    )
+
+    noise_scheduler = DDPMScheduler(
+        beta_start=0.00085,
+        beta_end=0.012,
+        beta_schedule='scaled_linear',
+        num_train_timesteps=1000,
+        tensor_format='pt'
+    )
+
+    # load dataset
+
+    def collate_fn(examples):
+        pixel_values = torch.stack([example['pixel_values'] for example in examples if example is not None])
+        pixel_values.to(memory_format=torch.contiguous_format).float()
+        input_ids = [example['input_ids'] for example in examples if example is not None]
+        padded_tokens = tokenizer.pad({'input_ids': input_ids}, return_tensors='pt', padding=True)
+        return {
+            'pixel_values': pixel_values,
+            'input_ids': padded_tokens.input_ids,
+            'attention_mask': padded_tokens.attention_mask,
+        }
+
+    train_dataset = LocalBase(
+        args.dataset,
+        args.resolution,
+        args.ucg,
+        args.resize_interp,
+        args.resize,
+        args.shuffle,
+        tokenizer
+    )
+
+    train_dataloader = torch.utils.data.DataLoader(
+        train_dataset,
+        shuffle=args.shuffle,
+        batch_size=args.batch_size,
+        collate_fn=collate_fn,
+    )
+
+    lr_scheduler = get_scheduler(
+        'constant',
+        optimizer=optimizer
+    )
+
+    weight_dtype = torch.float16 if args.fp16 else torch.float32
+
+    # move models to device
+    vae = vae.to(device, dtype=weight_dtype)
+    unet = unet.to(device, dtype=torch.float32)
+    text_encoder = text_encoder.to(device, dtype=weight_dtype)
+
+    # create ema
+    if args.use_ema:
+        ema_unet = EMAModel(unet.parameters())
+    
+    print(get_gpu_ram())
+
+    num_steps_per_epoch = len(train_dataloader)
+    progress_bar = tqdm.tqdm(range(args.epochs * num_steps_per_epoch), desc="Total Steps", leave=False)
+    global_step = 0
+
+    def save_checkpoint():
+        if args.use_ema:
+            ema_unet.copy_to(unet.parameters())
+        pipeline = StableDiffusionPipeline(
+            text_encoder=text_encoder,
+            vae=vae,
+            unet=unet,
+            tokenizer=tokenizer,
+            scheduler=PNDMScheduler(
+                beta_start=0.00085, beta_end=0.012, beta_schedule="scaled_linear", skip_prk_steps=True
+            ),
+            safety_checker=StableDiffusionSafetyChecker.from_pretrained("CompVis/stable-diffusion-safety-checker"),
+            feature_extractor=CLIPFeatureExtractor.from_pretrained("openai/clip-vit-base-patch32"),
+        )
+        pipeline.save_pretrained(args.output_path)
+    
+    # train!
+    for epoch in range(args.epochs):
```

Comment:
You have a pretty large function/loop here.  Suggest breaking this out into a function outside of `main()` that perhaps accepts closured functions as arguments. Or have a dataclass object that contains state passed in?